src/probnum/diffeq/odesolver.py

```diff
@@ -3,18 +3,103 @@
 """
 
 from abc import ABC, abstractmethod
+import warnings
+
+from probnum.diffeq.odesolution import ODESolution
 
 
 class ODESolver(ABC):
     """
     Interface for ODESolver.
     """
 
-    @abstractmethod
-    def solve(self, ivp, minstep, maxstep, **kwargs):
+    def __init__(self, ivp):
+        self.ivp = ivp
+        self.num_steps = 0
+
+    def solve(self, firststep, steprule, **kwargs):
         """
-        Every ODE solver has a solve() method.
-        Optional: callback function. Allows  e.g. printing variables
-        at runtime.
+        Solve an IVP.
+
+        Parameters
+        ----------
+        firststep : float
+            First step for adaptive step size rule.
+        steprule : StepRule
+            Step-size selection rule, e.g. constant steps or adaptive steps.
+        """
+        t, current_rv = self.initialise()
+        times, rvs = [t], [current_rv]
+        stepsize = firststep
+
+        while t < self.ivp.tmax:
+
+            t_new = t + stepsize
+            proposed_rv, errorest = self.step(t, t_new, current_rv, **kwargs)
+
+            if steprule.is_accepted(stepsize, errorest):
+                self.num_steps += 1
+                self.pre_accepted_callback(
```

Comment:
hmmm good point. I thought this name was a good idea to indicate that it is *after* the decision to accept the step but *before* actually accepting it storage wise. Maybe even more explicit than `*_applystep_*` with e.g. `pre_store(_accepted)_step`?

Reply:
To be clear: I thought it was a good idea but now that you say it I dont think that anymore :)

Reply:
It feels like both namings are not completely ideal, but I'm also fine with both since I think it will be hard to 100% describe it through its variable name.

Is "pre-something" really necessary actually? The callback does have access neither to the storage (`times` and `rvs`), nor to anything that changes with `t=t_new` and `current_rv=proposed_rv`. As I see it, it could actually even be moved *after* these, to line 49? If that is indeed the case it would simplify the naming. Maybe even just `callback` is good enough, as long as it is documented somewhere (e.g. the docstring of `solve`) that the callback is triggered after each accepted step?

Reply:
I agree, it would probably cover 99% of the applications if it is moved to the end (of the if accepted block). Suggestion: calling it `method_callback`? This could distinguish it from a `user_callback`/`problem_callback` or sth which a user specifies. I'll draft a proposal.

Reply:
`method_callback` sounds good! Thanks for working on it :)